appsmithorg/appsmith
# feat: added column freeze and unfreeze functionality to table widget

## Description

**PRD**: https://www.notion.so/appsmith/Ability-to-freeze-columns-dd118f7ed2e14e008ee305056b79874a?d=300f4968889244da9f737e1bfd8c06dc#2ddaf28e10a0475cb69f1af77b938d0b

This PR adds the following features to the table widget:

- Freeze the columns to the left or right of the table.(Both canvas and page view mode).
- Unfreeze the frozen columns. (Both canvas and page view mode).
- Columns that are left frozen, will get unfrozen at a position after the last left frozen column. (Both canvas and page view mode).
- Columns that are right frozen, will get unfrozen at a position before the first right frozen column. (Both canvas and page view mode).
- Column order can be persisted in the Page view mode.
- Users can also unfreeze the columns that are frozen by the developers.
- Columns that are frozen cannot be reordered(Both canvas and page view mode)
- **Property pane changes (Columns property)**:
     - If the column is frozen to the left then that column should appear at top of the list.
     - If the column is frozen to the right then that column should appear at the bottom of the list.
     - The columns that are frozen cannot be moved or re-ordered in the list. They remain fixed in their position.
- In-Page mode, If there is a change in frozen or unfrozen columns in multiple tables then the order of columns and frozen and unfrozen columns should get persisted on refresh i.e. changes should get persisted across refreshes.

Fixes #3265 #18230

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

- Manual. Below are the cases that were tested manually:
    - Existing frozen columns should be frozen when server-side pagination is enabled.
    - Two applications having two different tables should persist their column order and frozen items.
    - Scroll bars are appearing as expected.
    - Edit mode functionalities are working as expected.
    - Hidden columns should appear before the first right frozen column but should be overlayed by the right frozen column in such a way that the column is hidden.
    - Custom columns that are added should appear as before the first right frozen column.
- Jest: Added test for `generateNewColumnOrderFromStickyValue`
- Cypress: Added cypress test: `column_freeze_spec.js`
     - Check column freeze and unfreeze mechanism in canavs mode
         - Column freeze and unfreeze testing via propertypane
         - Column freeze and unfreeze testing via dropdown
    - Check column freeze and unfreeze mechanism in page mode
        - Column freeze and unfreeze testing with 0 pre-frozen columns
        - Column freeze and unfreeze testing with multiple pre-frozen columns 

### Test Plan
> https://github.com/appsmithorg/TestSmith/issues/2164
https://appsmith.notion.site/appsmith/TEST-PLAN-Table-Widget-v2-2-Column-Freeze-0aa438ec5be94702a87a480fe2952c53

### Issues raised during DP testing
>https://github.com/appsmithorg/appsmith/pull/18757#issuecomment-1366457654
https://github.com/appsmithorg/appsmith/pull/18757#issuecomment-1384070876
https://github.com/appsmithorg/appsmith/pull/18757#issuecomment-1366580010
https://github.com/appsmithorg/appsmith/pull/18757#issuecomment-1385047976



## Checklist:
### Dev activity


### QA activity:
